src/Nethermind/Nethermind.Serialization.Rlp/RlpStream.cs

```diff
@@ -1302,57 +1303,107 @@ public ulong DecodeUlong()
             return bytes.Length == 0 ? 0L : bytes.ReadEthUInt64();
         }
 
-        public byte[] DecodeByteArray() => DecodeByteArraySpan().ToArray();
-
-        public ReadOnlySpan<byte> DecodeByteArraySpan()
+        public byte[] DecodeByteArray()
         {
-            int prefix = ReadByte();
-            if (prefix == 0)
+            var span = DecodeByteArraySpan();
+            if (span.Length == 0)
             {
-                return Bytes.ZeroByte.Span;
+                return Array.Empty<byte>();
             }
 
-            if (prefix < 128)
+            if (span.Length == 1)
             {
-                return new[] { (byte)prefix };
+                int value = span[0];
+                var arrays = SingleByteArrays;
+                if ((uint)value < (uint)arrays.Length)
+                {
+                    return arrays[value];
+                }
+            }
+
+            return span.ToArray();
+        }
+
+        public ReadOnlySpan<byte> DecodeByteArraySpan()
+        {
+            int prefix = ReadByte();
+            ReadOnlySpan<byte> span = RlpStream.SingleBytes;
```

Comment:
Why not reuse `SingleByteArrays`?

Reply:
More checks get added in the conversion of a section of an array to a span than a slice of an existing span